subscription: Show data about attached subscriptions

M4rtinK opened the pull request:
This basically updates the second part of the spoke that
is shown after a successful subscrioption attempt. Type
of authentication used, final system purpose data as
well as information about attached subscriptions is shown.

The code that creates the listbox delegates describing an
attached subscription is not very nice, as it needs to
imperatively construct the rather complex listbox delegate
using the GTK/Pango API.

Ideally this would be done be describing the delegate in
Glade and then just instantiating it per item and filling
it with data. Unfortunately there is zero information
available for dynamically creating widgets with Glade
and I was simply not able to get it working in this way.

I guess this might explain why some of the dynamically
created widgets we have in Anaconda are implemented in C.

Also add a CSS file we use to make the list box background
transparent to make the delegates look better, especially
in case there is just one or two of them.

pep8speaks commented:
Hello @M4rtinK! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pyanaconda/ui/gui/spokes/lib/subscription.py`](https://github.com/rhinstaller/anaconda/blob/b58a8eeac44cd374d6934d9dd3c112f1eb735dfe/pyanaconda/ui/gui/spokes/lib/subscription.py):

> [Line 24:1](https://github.com/rhinstaller/anaconda/blob/b58a8eeac44cd374d6934d9dd3c112f1eb735dfe/pyanaconda/ui/gui/spokes/lib/subscription.py#L24): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 32:1](https://github.com/rhinstaller/anaconda/blob/b58a8eeac44cd374d6934d9dd3c112f1eb735dfe/pyanaconda/ui/gui/spokes/lib/subscription.py#L32): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 150:22](https://github.com/rhinstaller/anaconda/blob/b58a8eeac44cd374d6934d9dd3c112f1eb735dfe/pyanaconda/ui/gui/spokes/lib/subscription.py#L150): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 151:22](https://github.com/rhinstaller/anaconda/blob/b58a8eeac44cd374d6934d9dd3c112f1eb735dfe/pyanaconda/ui/gui/spokes/lib/subscription.py#L151): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent

* In the file [`pyanaconda/ui/gui/spokes/subscription.py`](https://github.com/rhinstaller/anaconda/blob/b58a8eeac44cd374d6934d9dd3c112f1eb735dfe/pyanaconda/ui/gui/spokes/subscription.py):

> [Line 976:100](https://github.com/rhinstaller/anaconda/blob/b58a8eeac44cd374d6934d9dd3c112f1eb735dfe/pyanaconda/ui/gui/spokes/subscription.py#L976): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 99 characters)
> [Line 978:100](https://github.com/rhinstaller/anaconda/blob/b58a8eeac44cd374d6934d9dd3c112f1eb735dfe/pyanaconda/ui/gui/spokes/subscription.py#L978): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (111 > 99 characters)
> [Line 1024:100](https://github.com/rhinstaller/anaconda/blob/b58a8eeac44cd374d6934d9dd3c112f1eb735dfe/pyanaconda/ui/gui/spokes/subscription.py#L1024): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)

M4rtinK commented:
Not sure how to fix the PEP8 error in a clean way - maybe we have just not touched `gi.repository` importing code since we enabled the checker ?

poncovka commented:
> Not sure how to fix the PEP8 error in a clean way - maybe we have just not touched `gi.repository` importing code since we enabled the checker ?

You can ignore this error when it is caused by imports from `gi.repository`.

poncovka commented on pyanaconda/ui/gui/spokes/lib/subscription.py:
```diff
@@ -100,3 +108,135 @@ def fill_combobox(combobox, user_provided_value, valid_values):
     # set the active id (what item should be selected in the combobox)
     combobox.set_active_id(active_id)
+
+    def clear_listbox(listbox):
```
It looks like these functions are nested in the function `fill_combobox`. Is it necessary?

M4rtinK replied:
Yeah, if it'S used just in a single place, we can drop the function and just include it.

poncovka replied:
That is not necessary, I'm just not sure how it works. The functions `add_attached_subscription_delegate` and `populate_attached_subscriptions_listbox` are nested in the function `fill_combobox`, but `fill_combobox` doesn't call them.

M4rtinK replied:
Oh, now I  see it! Yeah, the indentation is totally wrong - but given that it was wrong for all the functions I've added, I have completely missed that. :P Should be fixed now.

poncovka commented on pyanaconda/ui/gui/spokes/subscription.css:
```diff
@@ -0,0 +1,4 @@
+.subscriptions_listbox {
```
How does it work? I couldn't find the code that loads this file.

M4rtinK replied:
It should bind to the GtkListBox named "subscriptions_listbox" in glade:
https://github.com/rhinstaller/anaconda/blob/master/pyanaconda/ui/gui/spokes/subscription.glade#L941

I *think* the dot in the front makes it bind to the a name, while other prefixes make ti bind to something else (I think I saw #, which does something else than the dot).

In any case, this should make the listbox invisible by itself, so it should not look bad in the usual case of systems having just a couple attached subscriptions yet it will provided the needed scrolling in cases there are more subscription delegates then what can fit on the screen.

VladimirSlavik replied:
I think the problem was, "How is this file even loaded"?

M4rtinK replied:
The answer is simple - it is not loaded, at all! GAH, damn you Glade & GTK! :P

 Turns out this got solved by including the CSS definitions in the main `anaconda-gtk.css` file and this was just some local leftover from an earlier attempt back then to have these definitions separately. 

Anyway, scrapping this file & putting the definitions to `anaconda-gtk.css` - that should be the correct state.

M4rtinK commented on pyanaconda/ui/gui/spokes/subscription.py:
```diff
@@ -970,7 +971,64 @@ def _update_subscription_state(self):
+        self._sla_status_label.set_text(final_sla_string)
+
+        # usage
+        final_usage_string = ("Usage: {}").format(self.system_purpose_data.usage)
```
Indeed! There is even the `()` but not the underscore... On the bright sight, that's one more "fixed by rebase" bug. ;-)

Should be fixed now.

M4rtinK commented:
PR updated:
* fixed the indentation issues with the attached subscription listbox filling code
* added sane CSS definition
* added missing string translation

poncovka commented:
@M4rtinK, I think you didn't push the changes to the branch :-)

M4rtinK commented:
@poncovka Gah! I have really been looking into this for far too long. ;-) Should be finally up to date now. :)

M4rtinK commented:
jenkins, test this please

M4rtinK commented:
Even though GitHub shows "in progress" for the CI, when looking to Jenkins directly, both test runs have finished with success. So this should be clear for merging. :)
